src/Algebra/Solver/CommutativeMonoid/Normal.agda

```diff
@@ -0,0 +1,140 @@
+------------------------------------------------------------------------
+-- The Agda standard library
+--
+-- Normal forms in commutative monoids
+--
+-- Adapted from Algebra.Solver.Monoid.Normal
+------------------------------------------------------------------------
+
+{-# OPTIONS --cubical-compatible --safe #-}
+
+open import Algebra.Bundles using (CommutativeMonoid)
+
+module Algebra.Solver.CommutativeMonoid.Normal {c ℓ} (M : CommutativeMonoid c ℓ) where
+
+import Algebra.Properties.CommutativeSemigroup as CSProperties
+import Algebra.Properties.Monoid.Mult as MultProperties
+open import Data.Fin.Base using (Fin; zero; suc)
+open import Data.Nat as ℕ using (ℕ; zero; suc; _+_)
+open import Data.Vec.Base using (Vec; []; _∷_; lookup; replicate; zipWith)
+import Data.Vec.Relation.Binary.Pointwise.Inductive as Pointwise
+open import Relation.Binary.Definitions using (DecidableEquality)
+import Relation.Binary.Reasoning.Setoid as ≈-Reasoning
+import Relation.Nullary.Decidable as Dec
+
+open CommutativeMonoid M
+open MultProperties monoid using (_×_; ×-homo-1; ×-homo-+)
+open CSProperties commutativeSemigroup using (interchange)
+open ≈-Reasoning setoid
+
+private
+  variable
+    n : ℕ
+
+
+------------------------------------------------------------------------
+-- Monoid expressions
+
+open import Algebra.Solver.Monoid.Expression rawMonoid public
+  hiding (NormalAPI)
+
+------------------------------------------------------------------------
+-- Normal forms
+
+-- A normal form is a vector of multiplicities (a bag).
+
+Normal : ℕ → Set
+Normal n = Vec ℕ n
+
+-- The semantics of a normal form.
+
+⟦_⟧⇓ : Normal n → Env n → Carrier
+⟦ []    ⟧⇓ _       = ε
+⟦ n ∷ v ⟧⇓ (a ∷ ρ) = (n × a) ∙ (⟦ v ⟧⇓ ρ)
+
+-- We can decide if two normal forms are /syntactically/ equal.
+
+infix 5 _≟_
+
+_≟_ : DecidableEquality (Normal n)
+nf₁ ≟ nf₂ = Dec.map Pointwise-≡↔≡ (decidable ℕ._≟_ nf₁ nf₂)
+  where open Pointwise using (Pointwise-≡↔≡; decidable)
+
+------------------------------------------------------------------------
+-- Constructions on normal forms
+
+-- The empty bag.
+
+empty : Normal n
+empty = replicate _ 0
+
+-- A singleton bag.
+
+sg : (i : Fin n) → Normal n
```

Comment:
Could we take this opportunity to deprecate this cryptic name and replace it with `singleton`?